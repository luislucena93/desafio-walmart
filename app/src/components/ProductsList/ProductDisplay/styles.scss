@import '../../../commons/styleVars';

.productDisplay {
  width: 20%;
  min-width: 220px;
  margin: 20px 10px;
  background-color: white;
  border-radius: 10px;
  padding: 10px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}

.thumbnail {
  width: 100%;
}

.separator {
  width: calc(100% + 20px);
  margin-left: -10px;
  border-bottom: 5px solid $backgroundColor;
}

.productDescription {
  color: gray;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  span {
    color: black;
    font-weight: 600;
  }

}

.price {
  color: #6e6e6e;
  font-weight: 600;
  font-size: 18px;
}

.addButton {
  background-color: white;
  -webkit-box-shadow: 1px 1px 6px 0px $shadowColor;
  -moz-box-shadow: 1px 1px 6px 0px $shadowColor;
  box-shadow: 1px 1px 6px 0px $shadowColor;
  border-radius: 10px;
  border: none;
  padding: 5px;
  margin: 0px 20px;
  color: $mainColor;
  font-weight: 600;
  font-size: 15px;
  &:focus {
    outline: none;
  }
  &:active {
    -webkit-box-shadow: 1px 1px 2px 0px $shadowColor;
    -moz-box-shadow: 1px 1px 2px 0px $shadowColor;
    box-shadow: 1px 1px 2px 0px $shadowColor;
  }
}

.addRemoveButtons {
  -webkit-box-shadow: 1px 1px 6px 0px $shadowColor;
  -moz-box-shadow: 1px 1px 6px 0px $shadowColor;
  box-shadow: 1px 1px 6px 0px $shadowColor;
  border-radius: 10px;
  background-color: $backgroundColor;
  color: $mainColor;
  font-weight: 600;
  font-size: 15px;
  display: flex;
  width: calc(100% - 40px);
  justify-content: center;
  align-items: center;
  margin: 0 20px;

  p {
    margin: 0;
    width: 100%;
    text-align: center;
  }
  button {
    background-color: white;
    color: $mainColor;
    font-weight: 600;
    font-size: 15px;
    padding: 5px;
    border: none;
    width: 40px;
    &:focus {
      outline: none;
    }
    &:active {
      -webkit-box-shadow: 1px 1px 2px 0px $shadowColor;
      -moz-box-shadow: 1px 1px 2px 0px $shadowColor;
      box-shadow: 1px 1px 2px 0px $shadowColor;
    }
  }
  :first-child {
    border-radius: 10px 0px 0px 10px;
  }
  :last-child {
    border-radius: 0px 10px 10px 0px;
  }
}