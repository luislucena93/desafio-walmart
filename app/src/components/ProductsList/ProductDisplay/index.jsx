import React from 'react';
import './styles.scss';
import FormatMoney from '../../../helpers/formatters';

const ProductDisplay = (props) => {
    const {product, addProduct, removeProduct, productQty} = props;

    const getButton = () => {
        if (productQty === 0) {
            return <button onClick={() => addProduct(product)} className={"addButton"}>Agregar</button>
        }
        return (
            <div className={"addRemoveButtons"}>
                <button onClick={() => removeProduct(product)} data-testid={"removeProduct"}>-</button>
                <p>{productQty}</p>
                <button onClick={() => addProduct(product)} data-testid={"addProduct"}>+</button>
            </div>);
    }

    return (
        <div className={"productDisplay"}>
            <div>
                <img className={'thumbnail'} src={`http://${product.image}`}/>
                <div className={"separator"}/>
                <p className={"productDescription"}><span>{product.brand}</span> {product.description}</p>
                <p className={"price"}>${FormatMoney(product.price)}</p>
            </div>
            {getButton()}
        </div>
    )
}

export default ProductDisplay;