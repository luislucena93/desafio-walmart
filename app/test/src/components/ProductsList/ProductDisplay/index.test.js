import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';
import ProductDisplay from "../../../../../../app/src/components/ProductsList/ProductDisplay/index";

describe('ProductDisplay component', () => {

    let mockAddProduct;
    let mockRemoveProduct;
    const testProduct = {
        id: 'testId',
        brand: 'testBrand',
        description: 'testDescription',
        image: 'testImage',
        price: 1000
    };

    beforeEach(() => {
        mockAddProduct = jest.fn();
        mockRemoveProduct = jest.fn();
    });

    it('should render ProdcutDisplay correctly for product with quantity 0', () => {
        const {container} = render(<ProductDisplay product={testProduct} productQty={0} addProduct={mockAddProduct}
                                                   removeProduct={mockRemoveProduct}/>);
        expect(container).toMatchSnapshot();
    });

    it('should render ProdcutDisplay correctly for product with quantity greater than 0', () => {
        const {container} = render(<ProductDisplay product={testProduct} productQty={2} addProduct={mockAddProduct}
                                                   removeProduct={mockRemoveProduct}/>);
        expect(container).toMatchSnapshot();
    });


    it('should call addProduct function when click the button "Agregar" for product with qty 0', () => {
        render(<ProductDisplay product={testProduct} productQty={0} addProduct={mockAddProduct}
                               removeProduct={mockRemoveProduct}/>);
        fireEvent.click(screen.getByText('Agregar'));
        expect(mockAddProduct).toHaveBeenCalledWith(testProduct);
    })

    it('should call addProduct function when click the button "+" for product with qty greater than 0', () => {
        render(<ProductDisplay product={testProduct} productQty={2} addProduct={mockAddProduct}
                               removeProduct={mockRemoveProduct}/>);
        fireEvent.click(screen.getByText('+'));
        expect(mockAddProduct).toHaveBeenCalledWith(testProduct);
    })

    it('should call deleteProduct function when click the button "-" for product with qty greater than 0', () => {
        render(<ProductDisplay product={testProduct} productQty={2} addProduct={mockAddProduct}
                               removeProduct={mockRemoveProduct}/>);
        fireEvent.click(screen.getByText('-'));
        expect(mockRemoveProduct).toHaveBeenCalledWith(testProduct);
    })

});