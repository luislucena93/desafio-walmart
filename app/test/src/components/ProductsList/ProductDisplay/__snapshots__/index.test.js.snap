// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDisplay component should render ProdcutDisplay correctly for product with quantity 0 1`] = `
<div>
  <div
    class="productDisplay"
  >
    <div>
      <img
        class="thumbnail"
        src="http://testImage"
      />
      <div
        class="separator"
      />
      <p
        class="productDescription"
      >
        <span>
          testBrand
        </span>
         
        testDescription
      </p>
      <p
        class="price"
      >
        $
        1.000
      </p>
    </div>
    <button
      class="addButton"
    >
      Agregar
    </button>
  </div>
</div>
`;

exports[`ProductDisplay component should render ProdcutDisplay correctly for product with quantity greater than 0 1`] = `
<div>
  <div
    class="productDisplay"
  >
    <div>
      <img
        class="thumbnail"
        src="http://testImage"
      />
      <div
        class="separator"
      />
      <p
        class="productDescription"
      >
        <span>
          testBrand
        </span>
         
        testDescription
      </p>
      <p
        class="price"
      >
        $
        1.000
      </p>
    </div>
    <div
      class="addRemoveButtons"
    >
      <button
        data-testid="removeProduct"
      >
        -
      </button>
      <p>
        2
      </p>
      <button
        data-testid="addProduct"
      >
        +
      </button>
    </div>
  </div>
</div>
`;
